src/Disks/IO/CachedOnDiskWriteBufferFromFile.cpp

```diff
@@ -56,69 +56,66 @@ bool FileSegmentRangeWriter::write(const char * data, size_t size, size_t offset
     if (finalized)
         return false;
 
+    if (expected_write_offset != offset)
+    {
+        throw Exception(
+            ErrorCodes::LOGICAL_ERROR,
+            "Cannot write file segment at offset {}, because expected write offset is: {}",
+            offset, expected_write_offset);
+    }
+
     auto & file_segments = file_segments_holder.file_segments;
 
-    if (current_file_segment_it == file_segments.end())
+    if (file_segments.empty() || file_segments.back()->isDownloaded())
     {
-        current_file_segment_it = allocateFileSegment(current_file_segment_write_offset, is_persistent);
+        allocateFileSegment(expected_write_offset, is_persistent);
     }
-    else
-    {
-        auto file_segment = *current_file_segment_it;
-        assert(file_segment->getCurrentWriteOffset() == current_file_segment_write_offset);
 
-        if (current_file_segment_write_offset != offset)
-        {
-            throw Exception(
-                ErrorCodes::LOGICAL_ERROR,
-                "Cannot write file segment at offset {}, because current write offset is: {}",
-                offset, current_file_segment_write_offset);
-        }
+    auto & file_segment = file_segments.back();
 
-        if (file_segment->range().size() == file_segment->getDownloadedSize())
+    while (size > 0)
```

Comment:
Maybe let's unit tests to cover the case that allocates several segments?

Reply:
Actually my test covers this case because I set a small `max_file_segment_size` in disk config for this test. But I can also add an assertion for that with a check in system.filesystem_cache_log where I can see how many file segments were cached during writing data on insert. Will do... So seems like we can do without a unit test.

Reply:
> But I can also add an assertion for that with a check in system.filesystem_cache_log

Not need to add after all because currently in test I check entries in this log, and each entry means a different file segment, so it is already asserted.